bidirectional virtual connection ids (#62)
tfpauly/quic-proxy: draft-pauly-masque-quic-proxy.md

- Target Connection ID: a QUIC Connection ID that is chosen by the target, and
is used in the Destination Connection ID field of packets from the client to
the target.
- Virtual Client Connection ID: a fake QUIC Connection ID that is chosen by the
client that the proxy MUST use when sending QUIC packets in forwarding mode.
- Virtual Target Connection ID: a fake QUIC Connection ID that is chosen by the
proxy that the client MUST use when sending QUIC packets in forwarding mode.

## Virtual Connection IDs

QUIC allows each endpoint of a connection to choose the connection IDs it

> receives with. Servers receiving QUIC packets may employ load balancing

Reviewer: ```suggestion
receives with. Servers receiving QUIC packets can employ load balancing
```

strategies such as those described in {{?QUIC-LB=I-D.ietf-quic-load-balancers}}

> which may choose to take advantage of this by encoding routing information in

Reviewer: ```suggestion
that encode routing information in
```

the connection ID. When operating in forwarding mode, clients send QUIC packets
destined for the Target directly to the Proxy. Since these packets are generated

> using the Target Connection ID, load balancers may not be able to route packets
> to the correct Proxy. The Virtual Target Connection ID is a connection ID chosen

Reviewer: ```suggestion
using the Target Connection ID, load balancers would not be able to route packets
to the correct Proxy if the packets were sent with the Target Connection ID.
The Virtual Target Connection ID is a connection ID chosen
```

the target.
- Virtual Target Connection ID: a fake QUIC Connection ID that is chosen by the
proxy that the client MUST use when sending QUIC packets in forwarding mode.

> - Virtual Client Connection ID: a fake QUIC Connection ID that is chosen by the

Reviewer: To match the order of the CIDs above, let's put Virtual CCID before Virtual TCID

client that the proxy MUST use when sending QUIC packets in forwarding mode.

## Virtual Connection IDs

> QUIC allows each endpoint of a connection to choose the connection IDs it

Reviewer 2: ```suggestion
QUIC allows each endpoint of a connection to choose the destination connection IDs it
```

> receives with. Servers receiving QUIC packets may employ load balancing

Reviewer 2: ```suggestion
receives. Servers receiving QUIC packets may employ load balancing
```

Author: > ... choose the destination connection IDs it receives

To me, that reads a little ambiguously. An endpoints of a QUIC connection "receives" connection IDs from its peer via NEW_CONNECTION_ID frames, but that's obviously not what you meant. That being said, I don't love the language as-is either, but it's a carry over from the previous PR. May or may not make sense to address it now.

Reviewer: Yeah, let's not address for now.

strategies such as those described in {{?QUIC-LB=I-D.ietf-quic-load-balancers}}
which may choose to take advantage of this by encoding routing information in
the connection ID. When operating in forwarding mode, clients send QUIC packets
destined for the Target directly to the Proxy. Since these packets are generated
using the Target Connection ID, load balancers may not be able to route packets
to the correct Proxy. The Virtual Target Connection ID is a connection ID chosen
by the Proxy that the Client uses when sending forwarded mode packets. The Proxy
replaces the Virtual Target Connection ID with the Target Connection ID prior to
forwarding the packet to the Target.

Similarly, clients supporting QUIC connection migration require that connection

> IDs aren't reused over multiple network paths to avoid linkability. The Virtual

Reviewer 2: I think this is true for all clients, can we just say that "QUIC requires that connections IDs aren't reused over multiple network paths to..."?

Author: I was thinking about differences between passive and active migration. In the passive case, the CID may not have changed despite having been reused over a new network path. I'm wondering if it is necessary to mention this here or if the language you propose is sufficient.

Reviewer 2: QUIC has text about that already, I'd be tempted to say that we just refer to that since it was very carefully worded 🤷

Author: thanks, updated the language and made a few other changes here: 65db5de9489c96463f48ab2d9d8edafdfc8f1d8f

Client Connection ID is a connection ID chosen by the Client that the Proxy uses
when sending forwarded mode packets. The Proxy replaces the Client Connection ID
with the Virtual Client Connection ID prior to forwarding the packet to the
Client. Clients take advantage of this to avoid linkability when migrating the

> client to proxy network path. Without the Virtual Client Connection ID, QUIC

Reviewer 2: To "a" proxy network path? To "the" proxy network path?

Author: good catch: 5c419f6be0ae9b855be73a397dc114a6085bb829

> packets destinted for the client would have the same connection ID.

Reviewer 2: ```suggestion
packets destined for the client would have the same connection ID.
```

requests for very short Connection IDs as conflicts, in anticipation of future
conflicts.

## Stateless Resets for Forwarded Mode QUIC Packets

While the lifecycle of forwarding rules are bound to the lifecycle of the
client<->proxy HTTP stream, a peer may not be aware that the stream has
terminated. If the above mappings are lost or removed without the peer's
knowledge, they may send forwarded mode packets even though the Client
or Proxy no longer has state for that connection. To allow the Client or
Proxy to reset the client<->target connection in the absence the mappings
above, a stateless reset token corresponding to the Virtual Connection ID

> may be provided.

Reviewer 3: This new text is great, can we also augment it with an example where one device reboots and loses state, and what happens in that scenario to properly close the connection?

Author: Thanks, I added an example in e191d7d958c2f127038f8df6426917c6958a763f. I went with a slightly different example, but I think it communicates the same thing.

Reviewer 3: Great, thanks!

requests for very short Connection IDs as conflicts, in anticipation of future
conflicts.

## Stateless Resets for Forwarded Mode QUIC Packets

While the lifecycle of forwarding rules are bound to the lifecycle of the
client<->proxy HTTP stream, a peer may not be aware that the stream has
terminated. If the above mappings are lost or removed without the peer's
knowledge, they may send forwarded mode packets even though the Client
or Proxy no longer has state for that connection. To allow the Client or
Proxy to reset the client<->target connection in the absence the mappings
above, a stateless reset token corresponding to the Virtual Connection ID

> may be provided.

Reviewer: ```suggestion
can be provided.
```

Consider a proxy that initiates closure of a client<->proxy QUIC connection.

> If the client is temporarily unresponsive or unreachable, the proxy may have

Reviewer: ```suggestion
If the client is temporarily unresponsive or unreachable, the proxy might have
```

considered the connection closed and removed all connection state (including

> the above stream mappings). If the client never learned about the closure, it

Reviewer: ```suggestion
the stream mappings used for forwarding). If the client never learned about the closure, it
```

might send forwarded mode packets to the proxy, assuming the stream mappings

> and client<->proxy connection are still intact. The proxy receives these
> forwarding mode packets, but doesn't have any state corresponding to the

Reviewer: ```suggestion
and client<->proxy connection are still intact. The proxy will receive these
forwarding mode packets, but won't have any state corresponding to the
```

length of the capsule. Note that in QUICv1, the length of the Connection ID
is limited to 20 bytes, but QUIC invariants allow up to 255 bytes.

> The REGISTER_CLIENT_CID and ACK_TARGET_CID capsule types includes a Virtual

Reviewer: ```suggestion
The REGISTER_CLIENT_CID and ACK_TARGET_CID capsule types include a Virtual
```

Connection ID
: A connection ID being acknowledged whose length is equal to Connection ID
Length. This is the real Target or Client Connection ID.

Virtual Connection ID Length
: The length of the virtual connection ID being provided. This must be a valid
connection ID length for the QUIC version used in the client<->proxy QUIC
connection. When forwarding mode is not negotiated, the length MUST be zero.
The Connection ID Length and Virtual Connection ID Length MAY differ in

> ACK_TARGET_CID capsules, but MUST be equal in REGISTER_CLIENT_CID capsules.

Reviewer 3: Why the discrepancy here? Is there a reason to force the lengths to be equal?

Author: The discrepancy is motivated by implementation considerations. If implementing forwarding in a way that requires fixing IP and/or UDP checksums, only length changes by a multiple of 4 can be done incrementally. If the length difference is not a multiple of 4, the entire checksum would have to be recomputed. One example of this constraint is documented in [Linux's BPF helpers](https://man7.org/linux/man-pages/man7/bpf-helpers.7.html) (bpf_csum_diff):

> Compute a checksum difference, from the raw buffer
                     pointed by from, of length from_size (that must be
                     a multiple of 4), towards the raw buffer pointed by
                     to, of size to_size (same remark).

In the Virtual Target Connection ID case, the requirement to support Virtual Connection IDs of a different length is quite clear. Load balancing strategies such as those described QUIC-LB may require a minimum CID length to encode information (e.g. 17 bytes in the [single-pass encryption](https://datatracker.ietf.org/doc/html/draft-ietf-quic-load-balancers#section-4.3.1) case). If the proxy intended on using an incremental checksum algorithm, it could make sure that the Virtual Target CID is a multiple of 4 larger than the Target CID. For example, if the Target CID is 15 bytes, it could provide a Virtual Target CID of 19 bytes (17 "real" bytes encoding/encrypting information followed by 2 randomly chosen bytes). The replace operation of a 19 bytes buffer to a 15 byte buffer would be compatible with incremental checksum algorithms since 4 bytes are removed.

Requiring the Target Connection ID be at least the length of the Virtual Target Connection ID would necessitate cooperation with the target and as such is a non-starter. The draft currently allows any type of length difference between the Target Connection ID and Virtual Target Connection ID. We could instead tighten this up by requiring that the Virtual Target Connection ID be at least as long as the Target Connection ID, but I'm not sure how useful this is.

By requiring that the Client CID and Virtual Client CID lengths be equal, we're guaranteeing that incremental checksum is possible. We could instead require that the difference in length between Client CID and Virtual Client CID be a multiple of 4, but that seems like too specific of a requirement - particularly if it is of no real consequence to simply require they be equal.

By taking the above Virtual Target Connection ID length difference case as a requirement, we're precluding the client from using incremental checksum algorithms in the first place because it must be able to translate between arbitrary lengths when sending. Given that the client already needs to handle arbitrary length differences and that Client and Virtual Client CIDs are BOTH generated by the client, I would think that it has sufficient flexibility in crafting a Virtual Client CID of equivalent length that meet its needs. If we chose not to require this, the inclusion of Virtual Client CID (this PR) would introduce the requirement to do a full checksum recalculation for the reasons described above.

Reviewer 3: Thanks for the detailed explanation. I would suggest making this a SHOULD instead of a MUST: if the absolute worst case scenario is recomputing the checksum on each packet, then that doesn't warrant a MUST. There might be use-cases where a client or proxy isn't in control of its CID length, and in such scenarios they'd rather pay the checksum cost instead of failing.

Author: Something I forgot to include that is also worth considering is that the Virtual Client CID must be at least as long as the Client CID to avoid linkability. Consider the case of connection migration. If the Virtual Client CID is 0xaabb and the Client CID is 0x11223344, and the Virtual Client CID was changed due to migration to 0xccdd, the packet would go from flags byte followed by 0xaabb3344 to 0xccdd3344. The "3344" would increase linkability.

The Virtual Target Connection ID has the same issue - the client<->proxy network path change would expose any extra client<->target CID bytes. I think in both cases the Virtual Connection ID must be at least as long as its counterpart.

Author: Addressed here: 6cad095ceea2b5d54233a5111253a2351a894149

Reviewer 3: Oh this doesn't match my understanding of how replacement works, which makes me wonder if the replacement is sufficiently well documented. I was assuming that in this scenario you would replace `0x11223344` with `0xaabb` (not `0xaabb3344`) so you might see these two packets in sequence:
* before replacement `0x11223344abcdef...` `0x11223344fedcba...`
* after replacement `0xaabbabcdef...` `0xaabbfedcba...`

Am I misunderstanding how replacement works?

Author: In a previous commit, the replacement where the virtual cid is smaller than the cid is described [here](https://github.com/tfpauly/quic-proxy/pull/62/commits/6cad095ceea2b5d54233a5111253a2351a894149#diff-4dd090ab6ff7820385219d3ec1e86133930ab3a3b71aee2fe6e1a61a459f8da4L472)

> If the Virtual Target Connection ID is smaller than the Target Connection ID,
 the client MUST only write the Virtual Target Connection ID bytes over the start
 of the Target Connection ID, leaving the remainder of the Target Connection ID
 unmodified.

That was removed in 6cad095ceea2b5d54233a5111253a2351a894149 when changing the requirement to be that the virtual cid be at least as long as the cid.

I can update to document each case (virtual smaller, equal, and larger) to try to make this clearer once we get some alignment on what (if any) requirements should exist and what replacement should look like.

I hadn't considered shrinking the packet when the virtual cid is smaller than the cid, but it would satisfy the exposure issue during migration and should work equally well. Assuming we go with this approach, the replacement is always a "full" replacement. @tfpauly not sure if you have any opinions here.

Author: @DavidSchinazi removed length restrictions in 02f2fd47be22fcfba66c5d085d9406bf3e752f4c

This length equality constraint is more easily satisfied by the client since it
controls both the Connection ID and Virtual Connection ID.

> Virtual Connection ID

Reviewer 3: Mentions of sender and receiver in this paragraph are unclear: I was expecting them to be the sender/receiver of the capsule being discussed, but that doesn't appear to be the case

Author: Agreed, how about something like this?

> The peer-chosen connection ID that the sender of forwarded mode packets MUST
> use when sending. The endpoint rewrites forwarding mode packets to contain the
> correct Connection ID prior to sending them.

Author: updated with the above language in 7a2ec8dfe7b644a8f93e903a2ef500d6d49972c5

> : The recevier-chosen connection ID that the sender MUST use when sending

Reviewer 3: ```suggestion
: The receiver-chosen connection ID that the sender MUST use when sending
```

packets in forwarding mode. The proxy rewrites forwarding mode packets to
contain the correct Connection ID prior to forwarding them.

Stateless Reset Token Length

> : The length of the stateless reset token that may be sent by the client or

Reviewer 3: This isn't specific to this PR, but it would be nice to add a paragraph explaining why proxies care about stateless reset tokens and in which scenarios this feature would be useful

Author: Great point, I added some language here: f6ab95a1c0313babf5eb4c29e8fad8485fb0cab6

Connection IDs of different lengths than the corresponding Target Connection
IDs.

When forwarding, the proxy sends a QUIC packet with the Virtual Client Target

> Connection ID in the QUIC short header, using the same socket between client

Reviewer 2: Same socket? Same 5-tuple?

Author: I'm reusing the language as above: https://github.com/tfpauly/quic-proxy/blob/050afbdca7d33c327b67a56d6a9ffe5a0353b988/draft-pauly-masque-quic-proxy.md?plain=1#L453 `git blame` shows @DavidSchinazi so I'll defer to him

Reviewer 3: Sockets are explicitly defined as UDP 4-tuples in Section 1.2

Once a client has sent "Proxy-QUIC-Forwarding" with a value of "?1", it MUST be
prepared to receive forwarded short header packets on the socket between itself
and the proxy for any Virtual Client Connection ID that it has registered with a
REGISTER_CLIENT_CID capsule. The client uses the Destination Connection ID field
of the received packet to determine if the packet was originated by the proxy,
or merely forwarded from the target. The client replaces the Virtual Client

> Connection ID with the real Client Connection ID.

Reviewer: ```suggestion
Connection ID with the real Client Connection ID before processing the packet further.
```
